src/wp-includes/media.php

```diff
@@ -2117,6 +2117,13 @@ function wp_img_tag_add_width_and_height_attr( $image, $context, $attachment_id
 		$size_array = wp_image_src_get_dimensions( $image_src, $image_meta, $attachment_id );
 
 		if ( $size_array ) {
+			// If the width is enforced through style (e.g. in an inline image), calculate the dimension attributes.
+			$style_width = preg_match( '/style="width:\s*(\d+)px;"/', $image, $match_width ) ? (int) $match_width[1] : 0;
+			if ( $style_width ) {
+				$size_array[1] = (int) round( $size_array[1] * $style_width / $size_array[0] );
+				$size_array[0] = $style_width;
+			}
```

Comment:
What if an inline `height` style was supplied? Shouldn't that be honored?

Reply:
The most robust I can think of:

```php
$css = array();
if ( str_contains( $image, 'style=' ) ) {
	$p = new WP_HTML_Tag_Processor( $image );
	if ( $p->next_tag( array( 'tag_name' => 'IMG' ) ) ) {
		$style = $p->get_attribute( 'style' );
		if ( $style ) {
			foreach ( preg_split( '/\s*;\s*/', $style ) as $property_pair ) {
				$property_tuple = preg_split( '/\s*:\s*/', trim( $property_pair ), 2 );
				if ( count( $property_tuple ) === 2 ) {
					$css[ strtolower( $property_tuple[0] ) ] = $property_tuple[1];
				}
			}
		}
	}
}

$css_width = null;
if ( isset( $css['width'] ) && preg_match( '/px$/', $css['width'] ) ) {
	$css_width = (int) $css['width'];
}
$css_height = null;
if ( isset( $css['height'] ) && preg_match( '/px$/', $css['height'] ) ) {
	$css_height = (int) $css['height'];
}

if ( $css_width && $css_height ) {
	$size_array = array( $css_width, $css_height );
} else {
	$image_meta = wp_get_attachment_metadata( $attachment_id );
	$size_array = wp_image_src_get_dimensions( $image_src, $image_meta, $attachment_id );
	if ( $size_array ) {
		if ( $css_width ) {
			$size_array[1] = (int) round( $size_array[1] * $css_width / $size_array[0] );
			$size_array[0] = $css_width;
		} elseif ( $css_height ) {
			$size_array[0] = (int) round( $size_array[0] * $css_height / $size_array[1] );
			$size_array[1] = $css_height;
		}
	}
}
```

Reply:
While I'm open to make it more robust, your last solution, while the most robust, seems excessive to me.

I do believe we're primarily optimizing here for a quite specific use-case, and I'm not sure how much value there is in being more flexible. If a need comes up, we can always expand, but for now I'd rather go for a simple solution.

As soon as we provide more flexibility, we would also have to cater for a potential CSS `height` specification as in your code example, so I feel that makes things unnecessarily complex, which is why I opted for the very specific approach of supporting only width as provided for inline images.